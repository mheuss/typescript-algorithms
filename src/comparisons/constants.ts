export enum Comparisons {
  LessThan = -1,
  Equal,
  GreaterThan,
}

export type ComparisonFunction<T> = (a: T, b: T) => Comparisons;
