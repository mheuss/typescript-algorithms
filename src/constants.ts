export enum ErrorCodes {
  COMPARISON_FUNCTION_REQUIRED = 1,
  LIST_IS_BROKEN_INTERNALLY,
  OPERATION_BEYOND_BOUNDS,
  FAILED_TO_SET,
  COMPARISON_ERROR,
  INVALID_TREE_DATA_FILE,
}
