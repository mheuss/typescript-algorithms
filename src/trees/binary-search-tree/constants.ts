export enum Sides {
  Left,
  Right,
}
