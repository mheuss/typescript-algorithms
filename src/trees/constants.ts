export interface ITreeObject<T> {
  children: Array<ITreeObject<T>>;
  value: T;
}

export enum Traversal {
  Breadth_First,
  DepthFirstInOrder,
  DepthFirstPreOrder,
  DepthFirstPostOrder,
}
