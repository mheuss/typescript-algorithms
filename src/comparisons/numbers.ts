import { ErrorCodes } from '../constants';
import { Comparisons } from './constants';

/**
 * Simple numeric comparison function.
 * @param {number} a
 * @param {number} b
 * @throws {ErrorCodes.COMPARISON_ERROR} If it can't compare two numbers.
 * @returns {Comparisons}
 */
export function numericComparison(a: number, b: number): Comparisons {
  if (a > b) {
    return Comparisons.GreaterThan;
  } else if (a < b) {
    return Comparisons.LessThan;
  } else if (a === b) {
    return Comparisons.Equal;
  }

  throw {
    code: ErrorCodes.COMPARISON_ERROR,
    message: `I can not compare ${a} and ${b}`,
  };
}
