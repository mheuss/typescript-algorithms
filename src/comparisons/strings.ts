import { ErrorCodes } from '../constants';
import { Comparisons } from './constants';

/**
 * Simple string comparison function that makes use of string.localeCompare
 * @param {string} a
 * @param {string} b
 * @returns {Comparisons}
 * @throws {ErrorCodes.COMPARISON_ERROR}
 */
export function stringComparison(a: string, b: string): Comparisons {
  try {
    return a.localeCompare(b);
  } catch (e) {
    throw {
      code: ErrorCodes.COMPARISON_ERROR,
      message: `I can not compare ${a} and ${b}`,
    };
  }
}
