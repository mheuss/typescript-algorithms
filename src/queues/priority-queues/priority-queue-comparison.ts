import { Comparisons } from '../../comparisons/constants';
import { ErrorCodes } from '../../constants';
import { PriorityQueueNode } from './priority-queue-node';

export function priorityQueueComparison<T>(
  a: PriorityQueueNode<T>,
  b: PriorityQueueNode<T>
) {
  if (a.priority > b.priority) {
    return Comparisons.GreaterThan;
  } else if (a.priority < b.priority) {
    return Comparisons.LessThan;
  } else if (a.priority === b.priority) {
    return Comparisons.Equal;
  }

  throw {
    code: ErrorCodes.COMPARISON_ERROR,
    message: `I can not compare ${a} and ${b}`,
  };
}
