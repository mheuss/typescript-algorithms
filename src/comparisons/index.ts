import { ErrorCodes } from '../constants';
import { ComparisonFunction } from './constants';
import { numericComparison } from './numbers';
import { stringComparison } from './strings';

/**
 * This function is used commonly to return the correct comparison function for a type of data.
 * @param incomingArray Array of data we will be looking at
 * @param comparison Function, if provided, that'll be used for comparing.
 * @throws If there is not a comparison function provided AND we are not dealing with numbers or string
 * @constructor
 */
export function SelectComparisonFunction<T>(
  incomingArray: T[],
  comparison?: ComparisonFunction<T>
): ComparisonFunction<T> {
  const sortMechanismToBePassedIntoHelper = comparison;

  if (sortMechanismToBePassedIntoHelper === undefined) {
    if (
      typeof incomingArray[0] === 'string' ||
      incomingArray[0] instanceof String
    ) {
      // @ts-ignore
      return stringComparison;
    } else if (typeof incomingArray[0] === 'number') {
      // @ts-ignore
      return numericComparison;
    } else {
      throw {
        code: ErrorCodes.COMPARISON_FUNCTION_REQUIRED,
        message:
          'A custom comparison function is needed for arrays of types other than string or numbers',
      };
    }
  } else {
    return sortMechanismToBePassedIntoHelper;
  }
}
