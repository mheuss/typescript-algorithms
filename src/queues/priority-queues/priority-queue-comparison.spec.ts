import { Comparisons } from '../../comparisons/constants';
import { priorityQueueComparison } from './priority-queue-comparison';

describe('Priority Queue Comparison', () => {
  it('Should correctly identify a greater than', () => {
    expect(
      priorityQueueComparison(
        { value: 'a', priority: 5 },
        {
          priority: 1,
          value: 'b',
        }
      )
    ).toEqual(Comparisons.GreaterThan);
  });

  it('Should correctly identify a left than', () => {
    expect(
      priorityQueueComparison(
        { value: 'a', priority: 5 },
        {
          priority: 10,
          value: 'b',
        }
      )
    ).toEqual(Comparisons.LessThan);
  });

  it('Should correctly identify an equals', () => {
    expect(
      priorityQueueComparison(
        { value: 'a', priority: 5 },
        { value: 'b', priority: 5 }
      )
    ).toEqual(Comparisons.Equal);
  });

  it('Should throw if I provide it garbage', () => {
    expect(() => {
      priorityQueueComparison({ value: 'a', priority: 5 }, {
        banana: 10,
        value: 'b',
      } as any);
    }).toThrowErrorMatchingSnapshot();
  });
});
