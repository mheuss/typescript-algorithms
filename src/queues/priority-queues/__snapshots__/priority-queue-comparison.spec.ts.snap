// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Priority Queue Comparison Should throw if I provide it garbage 1`] = `"I can not compare [object Object] and [object Object]"`;
